src/prank_trivial_parallel.cpp

```diff
@@ -0,0 +1,147 @@
+    alpha = atof(argv[3]);
+    input_file = argv[4];
+
+
+
+    //PHASE 1 : Read graph from input file
+
+    //print input file name follwoed by num_itera and alpha
+    cout << "Parallel parameters: " << input_file << " " << num_iters << " " << alpha << endl;
+
+    auto tStart = std::chrono::high_resolution_clock::now();
+    int num_pages = 0;
+    unordered_map <int, Page> pages; //mapping page ID to its struct
+
+    ifstream cin(input_file);
+
+    int id, out;
+    while(cin>>id) {
+        cin >> out;
+        if(!pages.count(id)) {
+            pages[id] = Page();
+            num_pages++;
+        }
+        if(!pages.count(out)) {
+            pages[out] = Page();
+            num_pages++;
+        }
+        pages[out].indegree++;
+        pages[out].in_ids.push_back(id);
+        pages[id].outdegree++;
+    }
+
+    auto end = std::chrono::high_resolution_clock::now();
+    auto elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - tStart);
+    cout << "Load time: " << elapsed.count()*1e-9 <<endl;
+    //cout<<"Loaded "<<num_pages<<" pages"<<endl;
+
+
+    // initialize pageRank
+    float equal_prob = 1.0 / num_pages;
+    for(auto& x : pages)
+        x.second.score = equal_prob;
+
+    float broadcastScore;
+
+    //omp_set_num_threads(num_threads);
+    //PHASE 2 : Page ID Distribution
+    //TODO: ID distribution logic for different threads comes here.
+
+
+    //PHASE 3 : iteratively update ranks
+    //Once page distribution is done, we can spawn the threads
+    //The code below simply distributes the pages across threads based on page id, we need to change it
+    tStart = std::chrono::high_resolution_clock::now();
+    omp_set_num_threads(num_threads);
+    for(int j=0;j<num_iters;j++){
+        clock_t tStart = clock();
+        broadcastScore = 0.0;
+
+        // evaluate the score for each node
+        #pragma omp parallel for reduction(+: broadcastScore) schedule (dynamic, 32)
+        for(int i=0;i<pages.size();i++){
+
+            // find the ID corresponding to the i-page
+            int idx = i;
+
+            // used to store the new score
+            pages[idx].score_new = 0.0;
+
+            // if the node has no outgoing edges, then add its value to the broadcast score
+            if(!pages[idx].outdegree)
+                broadcastScore += pages[idx].score;
+
+            // iterate over all the vertices with an incoming edge to compute the new value of the vertices
+            if(pages[idx].indegree>0){
+                for(vector <int> :: iterator it = pages[idx].in_ids.begin(); it != pages[idx].in_ids.end(); ++it)
+                    pages[idx].score_new += pages[*it].score / pages[*it].outdegree;
+            }
+
+            // apply pageRank equation
+            pages[idx].score_new = alpha * pages[idx].score_new + (1.0 - alpha) / num_pages;
+        }
+
+        // update broadcast score
+        broadcastScore = alpha * broadcastScore / num_pages;
```

Comment:
Is there an equivalent for this in the spmv implementation? If not, will that cause us problems?

Reply:
I am not sure what precisely is the equivalent for this in spmv, I simply follwed the matrix multiplication rule mentioned in that paper. So we should be fine with that.

Reply:
I think this is significant and will cause problems. See the sink section [here](http://cs.brown.edu/courses/cs016/static/files/assignments/projects/GraphHelpSession.pdf)

Also a small example:
```
(base) nmerz@Nathans-MacBook-Pro pagerank % cat csr_test3
0 0
0 1
1 1
1 2
1 3
2 0
```
```
(base) nmerz@Nathans-MacBook-Pro pagerank % diff <(sort csr_test3_csr_parallel_ranks) <(sort csr_test3_serial_ranks)
1,4c1,4
< 0 0.2777376473
< 1 0.2389857173
< 2 0.1903816313
< 3 0.1903816313
---
> 0 0.3094615782
> 1 0.2662833242
> 2 0.2121275480
> 3 0.2121275480
```

Reply:
Got it, lemme check this one and get back